cmd/kubeadm/app/phases/certs/certs.go

```diff
@@ -653,6 +653,79 @@ func UsingExternalCA(cfg *kubeadmapi.MasterConfiguration) (bool, error) {
 	return true, nil
 }
 
+// ReadyForJoinAsMaster determines whether the local pki folder of a node is ready for
+// bootstrapping an additional master node.
+func ReadyForJoinAsMaster(cfg *kubeadmapi.MasterConfiguration) error {
+	// CA exists (if the user is relying on an external CA, absence of ca.key is tolerated)
+	caKeyPath := filepath.Join(cfg.CertificatesDir, kubeadmconstants.CAKeyName)
+	if _, err := os.Stat(caKeyPath); os.IsExist(err) {
+		if err := validateCACertAndKey(certKeyLocation{cfg.CertificatesDir, kubeadmconstants.CACertAndKeyBaseName, "", "CA"}); err != nil {
+			return err
+		}
+	} else {
+		if err := validateCACert(certKeyLocation{cfg.CertificatesDir, kubeadmconstants.CACertAndKeyBaseName, "", "CA"}); err != nil {
+			return err
+		}
+	}
+
+	// API server certificate exists and has the expected alt names (including the controlPlaneEndpoint and the advertiseAddress of the joining master)
+	if err := validateSignedCert(certKeyLocation{cfg.CertificatesDir, kubeadmconstants.CACertAndKeyBaseName, kubeadmconstants.APIServerCertAndKeyBaseName, "API server"}); err != nil {
+		return err
+	}
+	if err := validateAPIServerAltNames(cfg); err != nil {
+		return err
+	}
+
+	// API server kubelet client certificate exists
+	if err := validateSignedCert(certKeyLocation{cfg.CertificatesDir, kubeadmconstants.CACertAndKeyBaseName, kubeadmconstants.APIServerKubeletClientCertAndKeyBaseName, "API server kubelet client"}); err != nil {
+		return err
+	}
+
+	// service account signing key exists
+	if err := validatePrivatePublicKey(certKeyLocation{cfg.CertificatesDir, "", kubeadmconstants.ServiceAccountKeyBaseName, "service account"}); err != nil {
+		return err
+	}
+
+	// front proxy CA exists (if the user is relying on an external CA, absence of ca.key is tolerated)
+	frontProxyCAKeyPath := filepath.Join(cfg.CertificatesDir, kubeadmconstants.FrontProxyCACertAndKeyBaseName)
+	if _, err := os.Stat(frontProxyCAKeyPath); os.IsExist(err) {
+		if err := validateCACertAndKey(certKeyLocation{cfg.CertificatesDir, kubeadmconstants.FrontProxyCACertAndKeyBaseName, "", "front-proxy CA"}); err != nil {
+			return err
+		}
+	} else {
+		if err := validateCACert(certKeyLocation{cfg.CertificatesDir, kubeadmconstants.FrontProxyCACertAndKeyBaseName, "", "front-proxy CA"}); err != nil {
+			return err
+		}
+	}
+
+	// front proxy client cert exists
+	if err := validateSignedCert(certKeyLocation{cfg.CertificatesDir, kubeadmconstants.FrontProxyCACertAndKeyBaseName, kubeadmconstants.FrontProxyClientCertAndKeyBaseName, "front-proxy client"}); err != nil {
+		return err
+	}
+
+	// if external etcd, validate if certificates required for connecting API server to etcd exists.
+	// kubeadm does not executes further validations of those certificates
+	if len(cfg.Etcd.Endpoints) > 0 {
```

Comment:
probably cleaner if we return earlier like so:
```php
if len(cfg.Etcd.Endpoints) <= 0 {
	return nil
}

....

return nil		
```

haven't checked if `Endpoints` is an `uint` thus `<= 0` vs `== 0`.